ocsmesh/features/constraint.py

```diff
@@ -0,0 +1,214 @@
+
+        # pylint: disable=W1116
+        if isinstance(value_type, ConstraintValueType):
+            self._val_type = value_type
+        elif isinstance(value_type, str):
+            self._val_type = ConstraintValueType[value_type.upper()]
+        else:
+            raise ValueError("Invalid input for value type!")
+
+
+    @property
+    def type(self):
+        return type(self)
+
+
+    @property
+    def value_type(self):
+        return self._val_type
+
+    @property
+    def satisfies(self):
+
+        '''
+        The function to compare a value with the constraint value
+        and evaluate wether it satisfies the constraint
+        function's needs to receive values to check as first argument
+        '''
+
+        # pylint: disable=R1705
+        if self.value_type == ConstraintValueType.MIN:
+            return np.greater
+        elif self.value_type == ConstraintValueType.MAX:
+            return np.less
+
+        raise ValueError("Invalid value type for constraint!")
+
+    @property
+    def rate_sign(self):
+        # TODO: Put this method in a mixin
+
+        '''
+        Based on the value-type of the constraints, return a sign
+        indicating whether rate is for expansion or contraction of
+        size outside calculated zone
+        '''
+
+        # pylint: disable=R1705
+        if self.value_type == ConstraintValueType.MIN:
+            return -1
+        elif self.value_type == ConstraintValueType.MAX:
+            return +1
+
+        raise ValueError("Invalid value type for constraint!")
+
+
+    @abstractmethod
+    def apply(self):
+        pass
+
+
+    def _apply_rate(self, ref_values, values, locations, mask):
+
+        if not np.any(mask):
+            return values # TODO: COPY?
+
+        new_values = values.copy().ravel()
+        bound_values = ref_values.copy().ravel()
+        coords = locations.reshape(-1, 2)
+
+        if self._rate is None:
+            return values # TODO: COPY?
+
+        if len(coords) != len(new_values):
+            raise ValueError(
+                "Number of locations and values"
+                + f" don't match: {len(coords)} vs {len(new_values)}")
+
+        mask_r = mask.copy().ravel()
+        nomask_r = np.logical_not(mask_r)
+
+        points = coords[mask_r]
+        xy = coords[nomask_r]
+
+        tree = cKDTree(points)
+        near_dists, neighbors = tree.query(xy)
+        temp_values = new_values[mask_r][neighbors] * (
+                1 + near_dists * self._rate * self.rate_sign)
+
+        # NOTE: No bounds are applied for rate
+        mask2 = np.logical_not(self.satisfies(
+                     new_values[nomask_r], temp_values))
+        # Double indexing copies, we want to modify "new_values"
+        temp_values_2 = new_values[nomask_r]
+        temp_values_2[mask2] = temp_values[mask2]
+        new_values[nomask_r] = temp_values_2
+
+        new_values = new_values.reshape(values.shape)
+        return  new_values
+
+
+# TODO:
```

Comment:
During the next refactoring phase take care of creating Mixin classes instead of adding all methods to base class